[no-shadow] declaring a global causes no-shadow to error

Hey, I think I found what's either a bug or a confusing way that globals and ambient types are handled (potentially similarly to the `no-undef` issue [mentioned in the FAQs](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-get-errors-from-the-no-undef-rule-about-global-variables-not-being-defined-even-though-there-are-no-typescript-errors)). 

I'm currently migrating a large codebase to use TypeScript. We have a global `has()` function that we use to determine what environment we're in. It's handled by our build system, and is set as a read-only global in our existing ESLint config. 

When I try and declare an ambient type for this global, I've noticed that `@typescript-eslint/no-shadow` thinks that I'm shadowing it. Removing `globals: { has: false }` from the config fixes this problem. Is this expected behavior?

Thanks in advance, and sorry if I'm overlooking something.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.

Please consider creating an isolated reproduction repo to make it easy for the volunteer maintainers debug your issue.
-->
I set up a fresh project by running the following in an empty directory:
```sh
yarn init -y
yarn add -D eslint typescript @typescript-eslint/parser @typescript-eslint/eslint-plugin
```

I then added these two files
```js
// .eslintrc.js
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  plugins: [
    '@typescript-eslint',
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  globals: {
    has: false,
  },
  rules: {
    "no-shadow": "off",
    "@typescript-eslint/no-shadow": ["error", { builtinGlobals: true }],
  }
};
```


```TS
// index.d.ts
declare const has: (environment: "dev" | "prod" | "test") => boolean;
```

**Expected Result**

I'd expect `yarn eslint index.d.ts` to not error.

**Actual Result**

It logged the following error:
```
 1:15  error  'has' is already declared in the upper scope  @typescript-eslint/no-shadow
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.0` |
| `@typescript-eslint/parser`        | `4.4.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.10.0` |
| `node`                             | `12.15.0` |

--- a/packages/eslint-plugin/src/rules/no-shadow.ts
+++ b/packages/eslint-plugin/src/rules/no-shadow.ts
@@ -3,7 +3,7 @@ import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
 import { DefinitionType, ScopeType } from '@typescript-eslint/scope-manager';
 import { AST_NODE_TYPES, ASTUtils } from '@typescript-eslint/utils';
 
-import { createRule } from '../util';
+import { createRule, isDefinitionFile } from '../util';
 import { isTypeImport } from '../util/isTypeImport';
 
 export type MessageIds = 'noShadow' | 'noShadowGlobal';
@@ -568,6 +568,28 @@ export default createRule<Options, MessageIds>({
     }
 
     /**
+     * Checks if the initialization of a variable has the declare modifier in a
+     * definition file.
+     * @param variable The variable to check.
+     * @returns Whether or not the initialization of a variable has the declare
+     * modifier in a definition file.
+     */
+    function isDeclareInDTSFile(variable: TSESLint.Scope.Variable): boolean {
+      const fileName = context.filename;
+      if (!isDefinitionFile(fileName)) {
+        return false;
+      }
+      return variable.defs.some(def => {
+        return (
+          (def.type === DefinitionType.Variable && def.parent.declare) ||
+          (def.type === DefinitionType.ClassName && def.node.declare) ||
+          (def.type === DefinitionType.TSEnumName && def.node.declare) ||
+          (def.type === DefinitionType.TSModuleName && def.node.declare)
+        );
+      });
+    }
+
+    /**
      * Checks the current context for shadowed variables.
      * @param scope Fixme
      */
@@ -602,6 +624,11 @@ export default createRule<Options, MessageIds>({
 
         // ignore configured allowed names
         if (isAllowed(variable)) {
+          continue;
+        }
+
+        // ignore variables with the declare keyword in .d.ts files
+        if (isDeclareInDTSFile(variable)) {
           continue;
         }
 

--- a/packages/eslint-plugin/src/rules/no-shadow.ts
+++ b/packages/eslint-plugin/src/rules/no-shadow.ts
@@ -569,9 +569,6 @@ export default createRule<Options, MessageIds>({
     /**
      * Checks if the initialization of a variable has the declare modifier in a
      * definition file.
-     * @param variable The variable to check.
-     * @returns Whether or not the initialization of a variable has the declare
-     * modifier in a definition file.
      */
     function isDeclareInDTSFile(variable: TSESLint.Scope.Variable): boolean {
       const fileName = context.filename;

--- a/packages/eslint-plugin/tests/rules/no-shadow/no-shadow.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-shadow/no-shadow.test.ts
@@ -931,6 +931,48 @@ function foo<T extends (...args: any[]) => any>(fn: T, args: any[]) {}
         },
       ],
     },
+    {
+      code: `
+declare const has = (environment: 'dev' | 'prod' | 'test') => boolean;
+      `,
+      errors: [
+        {
+          data: {
+            name: 'has',
+          },
+          messageId: 'noShadowGlobal',
+        },
+      ],
+      languageOptions: {
+        globals: {
+          has: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare const has: (environment: 'dev' | 'prod' | 'test') => boolean;
+const fn = (has: string) => {};
+      `,
+      errors: [
+        {
+          data: {
+            name: 'has',
+            shadowedColumn: 15,
+            shadowedLine: 2,
+          },
+          messageId: 'noShadow',
+        },
+      ],
+      filename: 'foo.d.ts',
+      languageOptions: {
+        globals: {
+          has: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
   ],
   valid: [
     'function foo<T = (arg: any) => any>(arg: T) {}',
@@ -1478,6 +1520,135 @@ type A = 1;
 type A = 1;
       `,
       options: [{ hoist: 'functions' }],
+    },
+    {
+      code: `
+declare const foo1: boolean;
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          foo1: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare let foo2: boolean;
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          foo2: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare var foo3: boolean;
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          foo3: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+function foo4(name: string): void;
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          foo4: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare class Foopy1 {
+  name: string;
+}
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          Foopy1: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare interface Foopy2 {
+  name: string;
+}
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          Foopy2: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare type Foopy3 = {
+  x: number;
+};
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          Foopy3: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare enum Foopy4 {
+  x,
+}
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          Foopy4: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare namespace Foopy5 {}
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          Foopy5: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
+    },
+    {
+      code: `
+declare;
+foo5: boolean;
+      `,
+      filename: 'baz.d.ts',
+      languageOptions: {
+        globals: {
+          foo5: false,
+        },
+      },
+      options: [{ builtinGlobals: true }],
     },
   ],
 });

